lib/Parse/ParseExpr.cpp

```diff
@@ -2357,6 +2351,38 @@ DeclNameRef Parser::parseDeclNameRef(DeclNameLoc &loc,
   return DeclNameRef({ Context, baseName, argumentLabels });
 }
 
+ParserResult<Expr>
+Parser::parseMagicIdentifierLiteral(MagicIdentifierLiteralExpr::Kind Kind) {
+  SyntaxKind SKind = getMagicIdentifierSyntaxKind(Kind);
+
+  SyntaxParsingContext ExprContext(SyntaxContext, SyntaxContextKind::Expr);
+  ExprContext.setCreateSyntax(SKind);
+  SourceLoc Loc = consumeToken();
+
+  // If the magic identifier takes an argument list, capture it too.
+  ArgumentList *argList = nullptr; {
+    bool parseArgList = false;
+#define MAGIC_IDENTIFIER(NAME, STRING, SYNTAX_KIND)
+#define MAGIC_IDENTIFIER_TAKES_ARGUMENT(NAME) \
+  if (Kind == MagicIdentifierLiteralExpr::NAME) { parseArgList = true; }
```

Comment:
This blob could also be:

```c++
bool parseArgList = false;
switch (Kind) {
#define MAGIC_IDENTIFIER(NAME, STRING, SYNTAX_KIND)
#define MAGIC_IDENTIFIER_TAKES_ARGUMENT(NAME) \
case MagicIdentifierLiteralExpr::NAME:
#include "swift/AST/MagicIdentifierKinds.def"
  parseArgList = true;
default:
  parseArgList = false;
}
```

I suspect the exact same IL is generated either way.